// devdad-express-utils/dist/DatabaseConnection.js
import mongoose from "mongoose";
/**
 * Maximum number of connection retry attempts
 * Can be overridden with DB_MAX_RETRIES environment variable
 */
const MAX_RETRIES = parseInt(process.env.DB_MAX_RETRIES || '10');
/**
 * Initial interval between retry attempts in milliseconds
 * Can be overridden with DB_RETRY_INTERVAL environment variable
 */
const INITIAL_RETRY_INTERVAL = parseInt(process.env.DB_RETRY_INTERVAL || '3000'); // 3 seconds
/**
 * Maximum retry interval (caps exponential backoff)
 */
const MAX_RETRY_INTERVAL = 30000; // 30 seconds
/**
 * Database connection class for MongoDB using Mongoose.
 * Provides automatic reconnection, retry logic, and connection status monitoring.
 */
class DatabaseConnection {
    retryCount;
    isConnected;
    /**
     * Creates a new DatabaseConnection instance.
     * Sets up event listeners for connection events and application termination signals.
     */
    constructor() {
        this.retryCount = 0;
        this.isConnected = false;
        // Configure mongoose settings
        mongoose.set("strictQuery", true);
        // Handle connection events
        mongoose.connection.on("connected", () => {
            console.log("✅ MongoDB connected successfully");
            this.isConnected = true;
        });
        mongoose.connection.on("error", (err) => {
            console.error("❌ MongoDB connection error:", err);
            this.isConnected = false;
        });
        mongoose.connection.on("disconnected", () => {
            console.log("⚠️ MongoDB disconnected");
            this.isConnected = false;
            this.handleDisconnection();
        });
        // Handle application termination
        process.on("SIGINT", this.handleAppTermination.bind(this));
        process.on("SIGTERM", this.handleAppTermination.bind(this));
    }
    /**
     * Establishes a connection to MongoDB.
     * @throws {Error} If MONGO_URI is not defined in environment variables.
     */
    async connect() {
        try {
            if (!process.env.MONGO_URI) {
                throw new Error("MongoDB URI is not defined in environment variables");
            }
            const connectionOptions = {
                maxPoolSize: 10,
                serverSelectionTimeoutMS: 5000,
                socketTimeoutMS: 45000,
                family: 4, // Use IPv4
            };
            if (process.env.NODE_ENV === "development") {
                mongoose.set("debug", true);
            }
            await mongoose.connect(process.env.MONGO_URI, connectionOptions);
            this.retryCount = 0; // Reset retry count on successful connection
        }
        catch (error) {
            console.error("Failed to connect to MongoDB:", error.message);
            await this.handleConnectionError();
        }
    }
    /**
     * Handles connection errors by retrying the connection up to MAX_RETRIES times.
     * Uses exponential backoff for retry intervals.
     */
    async handleConnectionError() {
        if (this.retryCount < MAX_RETRIES) {
            this.retryCount++;
            // Calculate exponential backoff with jitter
            const exponentialDelay = Math.min(INITIAL_RETRY_INTERVAL * Math.pow(2, this.retryCount - 1), MAX_RETRY_INTERVAL);
            const jitter = Math.random() * 1000; // Add up to 1 second of random jitter
            const delay = exponentialDelay + jitter;
            console.log(`Retrying connection... Attempt ${this.retryCount} of ${MAX_RETRIES} (waiting ${Math.round(delay)}ms)`);
            await new Promise((resolve) => setTimeout(resolve, delay));
            return this.connect();
        }
        else {
            console.error(`Failed to connect to MongoDB after ${MAX_RETRIES} attempts. Please check:`, `1. Docker container is running: docker ps`, `2. MongoDB is accessible: ${process.env.MONGO_URI}`, `3. Network connectivity between containers`, `4. Environment variables are correctly set`);
            // Only exit in production, allow recovery in development
            if (process.env.NODE_ENV === 'production') {
                process.exit(1);
            }
            else {
                console.log('Development mode: keeping process alive for manual retry');
            }
        }
    }
    /**
     * Handles disconnection events by attempting to reconnect if not already connected.
     * Resets retry count for fresh reconnection attempts.
     */
    async handleDisconnection() {
        if (!this.isConnected) {
            console.log("Attempting to reconnect to MongoDB...");
            this.retryCount = 0; // Reset retry count for reconnection attempts
            await this.connect();
        }
    }
    /**
     * Handles application termination by closing the database connection gracefully.
     */
    async handleAppTermination() {
        try {
            await mongoose.connection.close();
            console.log("MongoDB connection closed through app termination");
            process.exit(0);
        }
        catch (err) {
            console.error("Error during database disconnection:", err);
            process.exit(1);
        }
    }
    /**
     * Gets the current connection status.
     * @returns An object containing connection status information.
     */
    getConnectionStatus() {
        return {
            isConnected: this.isConnected,
            readyState: mongoose.connection.readyState,
            host: mongoose.connection.host,
            name: mongoose.connection.name,
            retryCount: this.retryCount,
        };
    }
    /**
     * Manually reset retry count and attempt reconnection
     * Useful for recovering from Docker container restarts
     */
    async resetAndRetry() {
        console.log('Resetting connection state and retrying...');
        this.retryCount = 0;
        await this.connect();
    }
}
// Create a singleton instance
const dbConnection = new DatabaseConnection();
/**
 * Default export function to connect to MongoDB.
 * Establishes a connection with automatic retry logic and reconnection handling.
 *
 * @example
 * import connectDB from './DatabaseConnection';
 * await connectDB(); // Connects using MONGO_URI from environment
 */
export default dbConnection.connect.bind(dbConnection);
/**
 * Gets the current database connection status.
 *
 * @returns {Object} Connection status information including:
 * - isConnected: boolean - whether the database is connected
 * - readyState: number - mongoose connection ready state
 * - host: string - database host
 * - name: string - database name
 * - retryCount: number - current retry attempt count
 *
 * @example
 * import { getDBStatus } from './DatabaseConnection';
 * const status = getDBStatus();
 * console.log('DB connected:', status.isConnected);
 */
export const getDBStatus = dbConnection.getConnectionStatus.bind(dbConnection);
/**
 * Resets the connection state and attempts to reconnect.
 * Useful for recovering from Docker container restarts or network issues.
 *
 * @example
 * import { resetDBConnection } from './DatabaseConnection';
 * await resetDBConnection(); // Reset retry count and reconnect
 */
export const resetDBConnection = dbConnection.resetAndRetry.bind(dbConnection);
export { DatabaseConnection };
